### PR TITLE
Fixed #13296 -- Fixed ordering by Options.order_with_respect_to after deleting objects.

### PR DESCRIPTION
**Pull request by hramezani**
[ticket 13296](https://code.djangoproject.com/ticket/13296)
Original patch from @DrMeers

### CODE REVIEW

File: tests/order_with_respect_to/base_tests.py [hunk 1]
```diff
@@ -87,3 +87,23 @@ def test_recursive_ordering(self):
...
+            ['John', 'Paul', 'George', 'Ringo'],
+            attrgetter('text')
+        )
+        self.a1.delete()
+        self.a3.delete()
+        self.assertQuerysetEqual(
+            self.q1.answer_set.all(),
+            ['Paul', 'Ringo'],
+            attrgetter('text')
+        )
```
Comment (felixxm) [inline tests/order_with_respect_to/base_tests.py]: This assertion doesn't fail without a patch, IMO we can remove it.

File: django/db/models/base.py [hunk 2]
```diff
@@ -869,9 +869,10 @@ def _save_table(self, raw=False, cls=None, force_insert=False,
                 # autopopulate the _order field
                 field = meta.order_with_respect_to
                 filter_args = field.get_filter_kwargs_for_object(self)
-                order_value = cls._base_manager.using(using).filter(**filter_args).count()
-                self._order = order_value
-
+                max_order_value = cls._base_manager.using(using).filter(
+                    **filter_args
+                ).aggregate(Max('_order'))['_order__max']
+                self._order = max_order_value + 1 if max_order_value is not None else 0
```
Comment (felixxm) [inline django/db/models/base.py]: We could probably use more complicated expression to get rid of a temporary variable to make it more "atomic", e.g.
```python
self._order = cls._base_manager.using(using).filter(**filter_args).aggregate(
    max_order=Coalesce(
        ExpressionWrapper(Max('_order') + Value(1), output_field=IntegerField()),
        Value(0),
    ),
)['max_order']
```
:thinking:

File: tests/order_with_respect_to/base_tests.py [hunk 3]
```diff
@@ -87,3 +87,23 @@ def test_recursive_ordering(self):
...
         self.assertSequenceEqual(p1.get_post_order(), [p1_1.pk, p1_2.pk, p1_3.pk])
+
+    def test_deletion_and_insertion(self):
+        self.assertQuerysetEqual(
+            self.q1.answer_set.all(),
+            ['John', 'Paul', 'George', 'Ringo'],
+            attrgetter('text')
+        )
+        self.a1.delete()
+        self.a3.delete()
```
Comment (felixxm) [inline tests/order_with_respect_to/base_tests.py]: Can we move `a3` to another question instead of removing it?

**Comment by hramezani** [comment]
@felixxm, Thanks for the comments. I did your requested changes.

I think when we move answer to another question, the `_order` field of the answer will not update because of this condition:
https://github.com/django/django/blob/03fa846c6ab6abb7948ed6ecc84f35e7b6244b54/django/db/models/base.py#L866

So, if we have a test like this it will fail:
```
    def test_deletion_and_insertion(self):
        q2 = self.Question.objects.create(text="What is your name?")
        self.Answer.objects.create(text="Answer1", question=q2)
        self.Answer.objects.create(text="Answer2", question=q2)
        self.Answer.objects.create(text="Answer3", question=q2)
        self.Answer.objects.create(text="Answer4", question=q2)

        self.assertQuerysetEqual(
            self.q1.answer_set.all(),
            ['John', 'Paul', 'George', 'Ringo'],
            attrgetter('text')
        )
        self.a1.delete()
        self.a3.question = q2
        self.a3.save()
        self.assertQuerysetEqual(
            q2.answer_set.all(),
            ['Answer1', 'Answer2', 'Answer3', 'Answer4', 'Guido'],
            attrgetter('text')
        )
```

File: tests/order_with_respect_to/base_tests.py [hunk 4]
```diff
@@ -87,3 +87,20 @@ def test_recursive_ordering(self):
...
         self.assertSequenceEqual(p1.get_post_order(), [p1_1.pk, p1_2.pk, p1_3.pk])
+
+    def test_deletion_and_insertion(self):
+        q2 = self.Question.objects.create(text="What is your name?")
+        self.assertQuerysetEqual(
+            self.q1.answer_set.all(),
+            ['John', 'Paul', 'George', 'Ringo'],
+            attrgetter('text')
+        )
+        self.a1.delete()
```
Comment (claudep) [inline tests/order_with_respect_to/base_tests.py]: I don't think it's good practice to change instances created in `setUpTestData` as they should be identical for the whole test class.

**Comment by hramezani** [comment]
@felixxm I think we should find all objects after the deleted object and update their `_order` field of them with '-1'. this can be heavy if we have a lot of records or we do a lot of delete operations.
IMO, It does not worth.
BTW, I am open for any idea to add it.

**Comment by felixxm** [comment]
@hramezani Thanks for the update :+1: Agreed, the new logic is sufficient, let's leave it as it is.

**Pull request merged**